### PR TITLE
wazevo(amd64): initial implementation for `call` 

### PR DESCRIPTION
**Pull request by evacchi**
super early draft to understand if I am on the right track 🧪⚗️🐶 -("I don't know what I am doing")
nothing works, even regalloc blows up :P 

in practice, there is a lot of copy/paste going; obviously everything needs to be cleaned up (e.g. stuff needs to move to `instr_lower.go` like on arm64)
I enabled the e2e test "call" and working backwards from there. It turns out, quite a few things are needed 😬 

e.g. for `add` I introduced `aluRmiR: useKindOp1Op2` , `aluRmiR: defKindOp2`. I am not sure that is correct. Essentially `aluRmiR` needs 2 operands, and will overwrite one of the two.

EDIT meanwhile I figured that maybe @mathetake didn't mean to literally start from e2e testing but he was just pointing at the test case 😅  I think I can adapt `backend_test.go`

EDIT2: also I completely overlooked the clobberedRegs/spillSlots so that's likely the issue on e2e :D

### CODE REVIEW

**Comment by mathetake** [comment]
> EDIT meanwhile I figured that maybe @mathetake didn't mean to literally start from e2e testing but he was just pointing at the test case 😅 I think I can adapt backend_test.go

I meant e2e like this so looks good to me

> EDIT2: also I completely overlooked the clobberedRegs/spillSlots so that's likely the issue on e2e :D

you can ignore spill slots for now, and like-wise, you can just put TODO on args passed via stack

File: internal/engine/wazevo/backend/isa/amd64/machine.go [hunk 1]
```diff
@@ -132,11 +147,210 @@ func (m *machine) LowerInstr(instr *ssa.Instruction) {
...
+		ldr := m.allocateInstr()
+		switch r.Type {
+		case ssa.TypeI32, ssa.TypeI64:
+			ldr.asLEA(dst, reg)
+		case ssa.TypeF32, ssa.TypeF64, ssa.TypeV128:
+			panic("TODO") // ldr.asFpuLoad(operandNR(reg), amode, r.Type.Bits())
+		default:
+			panic("BUG")
+		}
+		m.insert(ldr)
```
Comment (mathetake) [inline internal/engine/wazevo/backend/isa/amd64/machine.go:322]: let's put the stack based args on hold. you can do it later.

```suggestion
		panic("TODO")
```

File: internal/engine/wazevo/backend/isa/amd64/machine.go [hunk 2]
```diff
@@ -132,11 +147,210 @@ func (m *machine) LowerInstr(instr *ssa.Instruction) {
...
+		m.InsertMove(arg.Reg, reg, arg.Type)
+	} else {
+		// TODO: we could use pair store if there's consecutive stores for the same type.
+		//
+		// Note that at this point, stack pointer is already adjusted.
+		bits := arg.Type.Bits()
+		am := m.resolveAddressModeForOffset(arg.Offset-slotBegin, bits, rspVReg)
+		store := m.allocateInstr()
+		store.asMovRM(reg, newOperandMem(am), bits/8)
+		m.insert(store)
```
Comment (mathetake) [inline internal/engine/wazevo/backend/isa/amd64/machine.go:302]: same. 
```suggestion
    panic("TODO")
```

**Review by mathetake — COMMENTED** [review]
I didn't know that call e2e test case needs ADD instruction, so could you create a more simple one so that you won't need to implement call and Add altogether? 

Also, feel free to put the panic TODO on the complex branches like stack args etc

File: internal/engine/wazevo/backend/isa/amd64/machine.go [hunk 3]
```diff
@@ -240,8 +327,19 @@ func (m *machine) Encode(context.Context) {
...
+		calleeFnOffset := refToBinaryOffset[r.FuncRef]
+		// calleeFnOffset points to the beginning of call target function.
+		// call is 5 bytes where the last 4 bytes represent the signed 32-bit offset. See the encoding of `call` instruction.
+		callInstrOffsetBytes := binary[instrOffset-1 : instrOffset+4]
+		diff := int64(calleeFnOffset) - (instrOffset)
+		// We backpatch in-place the relative value `diff`.
+		callInstr[1] = byte(diff)
+		callInstr[2] = byte(diff >> 8)
+		callInstr[3] = byte(diff >> 16)
+		callInstr[4] = byte(diff >> 24)
```
[the remark this answers is not in the document]
Comment (evacchi) [reply internal/engine/wazevo/backend/isa/amd64/machine.go]: huh good point

File: internal/engine/wazevo/backend/isa/amd64/abi_entry_preamble.go [hunk 4]
```diff
@@ -53,8 +57,22 @@ func (m *machine) compileEntryPreamble(sig *ssa.Signature) *instruction {
...
 	// 	The first two args are module context ptr and execution context ptr, and they are passed in %rax and %rcx.
-	if len(abi.Args) > 2 {
-		panic("TODO")
+	prReg := paramResultSlicePtr
+	if len(abi.Args) > 2 && len(abi.Rets) > 0 {
+		// paramResultSlicePtr is modified during the execution of goEntryPreamblePassArg,
+		// so copy it to another reg.
+		cur = m.move64(paramResultSliceCopied, paramResultSlicePtr, cur)
+		prReg = paramResultSliceCopied
+	}
```
Comment (mathetake) [inline internal/engine/wazevo/backend/isa/amd64/abi_entry_preamble.go:66]: q: is this the relevant change?

File: internal/engine/wazevo/backend/isa/amd64/abi_entry_preamble.go [hunk 5]
```diff
@@ -53,8 +57,22 @@ func (m *machine) compileEntryPreamble(sig *ssa.Signature) *instruction {
...
+
+	stackSlotSize := abi.AlignedArgResultStackSlotSize()
+	for i := range abi.Args {
+		if i < 2 {
+			// module context ptr and execution context ptr are passed in x0 and x1 by the Go assembly function.
+			continue
+		}
+		arg := &abi.Args[i]
+		cur = m.goEntryPreamblePassArg(cur, prReg, arg, -stackSlotSize)
```
Comment (mathetake) [inline internal/engine/wazevo/backend/isa/amd64/abi_entry_preamble.go:75]: same

**Review by mathetake — APPROVED** [review]

**Comment by mathetake** [comment]
let's do `ADD` instruction in a single function next, then after that, let's do call instruction with arguments

**Pull request merged**